File: BOJ/[9251] LCS/Baeksu/9251.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int N;
int dp[1001][1001];

int main() {
	string str1, str2;

	cin >> str1;
	cin >> str2;

	for (int i = 0; i < str1.size(); i++) {
 		for (int j = 0; j < str2.size(); j++) {
			if (str1[i] == str2[j])
				dp[i + 1][j + 1] = dp[i][j] + 1;
			else
				dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j]);
		}
	}
	cout << dp[str1.size()][str2.size()];
	return 0;
}

File: BOJ/[9251] LCS/DAEGUE/[9251] LCS/9251.py
import sys

def main():
    str1 = ' ' + sys.stdin.readline().rstrip()
    str2 = ' ' + sys.stdin.readline().rstrip()
    cal(str1, str2)


def cal(s1, s2):
    len1 = len(s1)
    len2 = len(s2)
    res = [[0] * len2 for j in range(len1)]
    for i in range(1, len1):
        for j in range(1, len2):
            if s1[i] == s2[j]:
                res[i][j] = res[i-1][j-1]+1
            else:
                res[i][j] = max(res[i-1][j], res[i][j-1])
    print(res[-1][-1])

main()